File: LeetCode/binary-tree-zigzag-level-order-traversal.cpp
// https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode *root)
    {
        vector<vector<int>> res;
        if (!root)
        {
            return res;
        }
        queue<TreeNode *> q;
        vector<int> currLevel;
        q.push(root);
        int level = -1;
        while (!q.empty())
        {
            level++;
            currLevel.clear();
            int nodeCount = q.size();
            while (nodeCount--)
            {
                auto currNode = q.front();
                q.pop();
                if (currNode->left)
                {
                    q.push(currNode->left);
                }
                if (currNode->right)
                {
                    q.push(currNode->right);
                }
                currLevel.push_back(currNode->val);
            }
            if (level % 2 != 0)
            {
                reverse(currLevel.begin(), currLevel.end());
            }
            res.push_back(currLevel);
        }
        return res;
    }
};

File: LeetCode/binary-tree-zigzag-level-order-traversal.java
// https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/

import java.util.ArrayList;
import java.util.List;
import java.util.Queue;
import java.util.LinkedList;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
  public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
    List<List<Integer>> result = new ArrayList<>();
    Queue<TreeNode> q = new LinkedList<>();

    if (root == null)
      return result;

    boolean isOrderLeft = true;
    q.offer(root);

    while (!q.isEmpty()) {
      // Clear the currLevel list for new level.
      LinkedList<Integer> currLevel = new LinkedList<>();
      int size = q.size();

      while (size-- > 0) {
        TreeNode curr = q.poll();

        if (curr.left != null)
          q.add(curr.left);
        if (curr.right != null)
          q.add(curr.right);

        if (isOrderLeft) {
          currLevel.addLast(curr.val);
        } else {
          currLevel.addFirst(curr.val);
        }
      }
      result.add(currLevel);
      isOrderLeft = !isOrderLeft;
    }

    return result;
  }
}